### PR TITLE
Jetty 9.4.x 7517 trie stack overflow

### LINKED ISSUE #7517: Some ArrayTrie methods throw StackOverflowError when cointaining a very large entry
**Jetty version(s)**
9.4.x

**Description**
`keySet()` and `toString()` recurse too deep on large entries, so they're likely to throw `StackOverflowError`.

Consider the following:

```java
Trie<String> trie = new ArrayTrie<>(Character.MAX_VALUE - 1);

char[] c1 = new char[Character.MAX_VALUE - 1];
Arrays.fill(c1, 'a');
String huge = new String(c1);
assertTrue(trie.put(huge, "wow"));
assertThat(trie.get(huge), is("wow"));

assertThat(trie.keySet(), notNullValue());   // <--- throws StackOverflowError
assertThat(trie.toString(), notNullValue()); // <--- throws StackOverflowError
```

### PR DESCRIPTION
**Pull request by gregw**
avoid recursion

### CODE REVIEW

**Comment by gregw** [comment]
Ooops this also includes the fix to #7518

**Review by lorban — CHANGES_REQUESTED** [review]
You broke `HttpParserTest` which means there's a lack of test coverage in `TrieTest`.

**Comment by gregw** [comment]
@lorban Added extra test. The problem was a bug with the return if the first non matching character was a big index one.

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 1]
```diff
@@ -331,6 +335,105 @@ public void testTrieOverflowReject(Function<Integer, Trie<String>> constructor)
...
+    {
+        Trie<String> trie = constructor.apply(Character.MAX_VALUE - 1);
+
+        char[] c1 = new char[Character.MAX_VALUE - 1];
+        Arrays.fill(c1, 'a');
+        String huge = new String(c1);
+        assertTrue(trie.put(huge, "wow"));
+        assertThat(trie.get(huge), is("wow"));
+
+        assertNotNull(trie.keySet());
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: Maybe assert that keyset contains `huge`?

File: jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java [hunk 2]
```diff
@@ -331,6 +335,105 @@ public void testTrieOverflowReject(Function<Integer, Trie<String>> constructor)
...
+        assertNotNull(trie.toString());
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/TrieTest.java]: Maybe assert that `toString()` contains the `huge` substring?

**Review by lorban — DISMISSED** [review]
I would slightly improve a couple of assertions, otherwise LGTM.

**Comment by yunussemreeee** [comment]
7528

**Pull request merged**